`NoSuchMethodError` for `com.fasterxml.jackson.module.kotlin.jacksonObjectMapper` after upgrading to 2.17.0

### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-module-kotlin/issues) and found nothing similar.
- [ ] I searched in the [issues of databind](https://github.com/FasterXML/jackson-databind/issues) and other modules used and found nothing similar.
- [ ] I have confirmed that the problem only occurs when using Kotlin.

### Describe the bug

Calling `jacksonObjectMapper()` after importing `com.fasterxml.jackson.module.kotlin.jacksonObjectMapper` after upgrading to 2.17.0 throws
```
Caused by: java.lang.NoSuchMethodError: 'com.fasterxml.jackson.databind.ObjectMapper com.fasterxml.jackson.module.kotlin.ExtensionsKt.jacksonObjectMapper$default(kotlin.jvm.functions.Function1, int, java.lang.Object)'
```

I believe that's related to https://github.com/FasterXML/jackson-module-kotlin/commit/0dbe80282dba1c99dbe1ab22981641bea2969148. By the looks of things there's `@JvmOverloads` and the new arg is defaulted so this should just work, shouldn't it?

FWIW, I've also tried `jacksonObjectMapper {}` but got 
```
Caused by: java.lang.NoSuchMethodError: 'com.fasterxml.jackson.databind.ObjectMapper com.fasterxml.jackson.module.kotlin.ExtensionsKt.jacksonObjectMapper(kotlin.jvm.functions.Function1)'
```


### To Reproduce

```kotlin
jacksonObjectMapper()
```


### Expected behavior

_No response_

### Versions

Kotlin: 1.9.20
Jackson-module-kotlin: 2.17.0
Jackson-databind: 2.17.0


### Additional context

Not sure if that matters at all but the method in question is called from a field initializer in a companion object of a Gradle plugin implemented in kotlin.

--- a/release-notes/CREDITS-2.x
+++ b/release-notes/CREDITS-2.x
@@ -15,7 +15,12 @@ Authors:
 
 Contributors:
 
-# 2.17.1 (not yet released)
+# 2.17.2 (not yet released)
+
+WrongWrong (@k163377)
+* #799: Fixed problem with code compiled with 2.17.x losing backward compatibility.
+
+# 2.17.1 (04-May-2024)
 
 WrongWrong (@k163377)
 * #776: Delete Duration conversion that was no longer needed

--- a/release-notes/VERSION-2.x
+++ b/release-notes/VERSION-2.x
@@ -16,6 +16,9 @@ Co-maintainers:
 === Releases ===
 ------------------------------------------------------------------------
 
+2.17.2 (not yet released)
+#799: Fixed problem with code compiled with 2.17.x losing backward compatibility.
+
 2.17.1 (04-May-2024)
 
 #776: Delete Duration conversion that was no longer needed.

--- a/src/main/kotlin/com/fasterxml/jackson/module/kotlin/Extensions.kt
+++ b/src/main/kotlin/com/fasterxml/jackson/module/kotlin/Extensions.kt
@@ -33,15 +33,17 @@ fun jsonMapper(initializer: JsonMapper.Builder.() -> Unit = {}): JsonMapper {
     return builder.build()
 }
 
-// region: JvmOverloads is set for bytecode compatibility for versions below 2.17.
-@JvmOverloads
+// region: Do not remove the default argument for functions that take a builder as an argument for compatibility.
+//   The default argument can be removed in 2.21 or later. See #775 for the history.
+fun jacksonObjectMapper(): ObjectMapper = jsonMapper { addModule(kotlinModule()) }
 fun jacksonObjectMapper(initializer: KotlinModule.Builder.() -> Unit = {}): ObjectMapper =
     jsonMapper { addModule(kotlinModule(initializer)) }
-@JvmOverloads
+
+fun jacksonMapperBuilder(): JsonMapper.Builder = JsonMapper.builder().addModule(kotlinModule())
 fun jacksonMapperBuilder(initializer: KotlinModule.Builder.() -> Unit = {}): JsonMapper.Builder =
     JsonMapper.builder().addModule(kotlinModule(initializer))
 
-@JvmOverloads
+fun ObjectMapper.registerKotlinModule(): ObjectMapper = this.registerModule(kotlinModule())
 fun ObjectMapper.registerKotlinModule(initializer: KotlinModule.Builder.() -> Unit = {}): ObjectMapper =
     this.registerModule(kotlinModule(initializer))
 // endregion

